# 실기 대비
# 실전 문제풀이
# set 1

## 1) 데이터 및 시나리오

### 시중 TV 제품별 분석

> 시중 판매 중인 TV정보를 온라인 쇼핑몰 웹페이지에서 크롤링 하여 분석을 하고자 한다. 해당 데이터를 기반으로 각 TV의 제원은 어떤 한지 제원에 따른 평가는 어떤 한지 알아보고 향후 신제품 출시에 참고하고자 한다.

### 데이터 개요

| 파일명 | 행 | 열 | 인코딩 |
|---|---:|---:|---|
| `TV.csv` | 666 | 11 | UTF-8 |

### 변수 상세

| 변수명 | 유형 | 설명 |
|---|---|---|
| `Product_Name` | string | 상품명 |
| `Stars` | float | 평가 평균 점수 |
| `Ratings` | int | 평가 수 |
| `Reviews` | int | 후기 수 |
| `current_price` | int | 현재 가격 |
| `MRP` | int | 공장 출고가 |
| `channel` | string | 프리미엄 채널 서비스 제공 목록 |
| `Operating_system` | string | 운영체제 |
| `Picture_quality` | string | 해상도 |
| `Speaker` | string | 스피커 |
| `Frequency` | string | 주사율 |

## 2) 문제

### 필요 라이브러리 함수 및 클래스 목록

| 목록 |
|---|
| `from sklearn.ensemble import RandomForestRegressor` |

In [1]:
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestRegressor

In [5]:
df = pd.read_csv('../../dataset/TV.csv')
#SCDI
display(df.shape)
display(df.columns)
display(df.dtypes)
display(df.isna().sum())

(666, 11)

Index(['Product_Name', 'Stars', 'Ratings', 'Reviews', 'current_price', 'MRP',
       'channel', 'Operating_system', 'Picture_quality', 'Speaker',
       'Frequency'],
      dtype='object')

Product_Name         object
Stars               float64
Ratings               int64
Reviews               int64
current_price         int64
MRP                   int64
channel              object
Operating_system     object
Picture_quality      object
Speaker              object
Frequency            object
dtype: object

Product_Name        0
Stars               0
Ratings             0
Reviews             0
current_price       0
MRP                 0
channel             0
Operating_system    0
Picture_quality     0
Speaker             0
Frequency           0
dtype: int64

### Q01.

주사율(`Frequency`)은 TV 제원 중 가격에 큰 영향을 미치는 제원이다.  
그런데 데이터 수집 중 실수가 발생하여 주사율(`Frequency`) 값 중 일부가 해상도(`Picture_quality`), 스피커(`Speaker`) 변수에 잘못 입력된 것이 발견되었다.  
세 변수에 들어간 주사율 데이터를 취합하여 전체 TV 중에서 주사율이 60Hz인 TV는 총 몇 대 인가?

※ 주사율은 2~3자리 숫자 뒤에 반드시 `Hz`라는 단위로 표기되어 있다.  
※ 세 변수 모두 주사율 정보가 없는 경우 결측치로 간주하고 이를 분석에서 제외하시오. `(정답 예시: 12)`

In [95]:
df_q1 = df.copy()
cols_q1 = ['Picture_quality', 'Speaker', 'Frequency']
display(df_q1[cols_q1]) 
"""
문제에는 Hz 만 보장하고 있음을 알 수 있다. 
freq에 Hz Refresh Rate 까지 있지만 아닐 수 있다는 뜻
그러니 문제에서 보장하는 것만 쓰도록 하자. 
"""
df_q1['pq-s-f'] = df_q1['Picture_quality'] + ' ' + df_q1['Speaker'] + ' ' + df_q1['Frequency']
#문자열 합칠 땐 공백구분 꼭하기
display(df_q1['pq-s-f'])

cond_q1 = df_q1['pq-s-f'].str.contains('Hz')
display(cond_q1.sum())
#isin은 값 자체가 있는지를 확인, str.contains는 문자열이 있는지를 확인
df_q1_cond = df_q1.loc[cond_q1, ].copy()
display(df_q1_cond['pq-s-f'].str.contains('60 Hz').sum())

# 개선
cond_q1_fix = df_q1['pq-s-f'].str.extract(r'(\d{2,3})\s*Hz')
display(type(cond_q1_fix), cond_q1_fix)
cond_q1_fix_60 = cond_q1_fix == '60' # 60은숫자고 '60'은 문자여서 조심해야함
display(cond_q1_fix_60)
display(cond_q1_fix_60.sum())


,Picture_quality,Speaker,Frequency
0,60 Hz Refresh Rate,2 x HDMI | 2 x USB,1 Year Warranty
1,HD Ready 1366 x 768 Pixels,20 W Speaker Output,60 Hz Refresh Rate
2,HD Ready 1366 x 768 Pixels,10 W Speaker Output,50 Hz Refresh Rate
3,HD Ready 1366 x 768 Pixels,20 W Speaker Output,60 Hz Refresh Rate
4,HD Ready 1366 x 768 Pixels,20 W Speaker Output,60 Hz Refresh Rate
...,...,...,...
661,60 Hz Refresh Rate,2 x HDMI | 1 x USB,1 YEAR
662,Ultra HD (4K) 3840 x 2160 Pixels,20 W Speaker Output,50 Hz Refresh Rate
663,60 Hz Refresh Rate,2 x HDMI | 2 x USB,3 Years Warranty
664,16 Speaker Output,50 Hz Refresh Rate,2 x HDMI | 2 x USB


0      60 Hz Refresh Rate 2 x HDMI | 2 x USB 1 Year W...
1      HD Ready 1366 x 768 Pixels 20 W Speaker Output...
2      HD Ready 1366 x 768 Pixels 10 W Speaker Output...
3      HD Ready 1366 x 768 Pixels 20 W Speaker Output...
4      HD Ready 1366 x 768 Pixels 20 W Speaker Output...
                             ...                        
661         60 Hz Refresh Rate 2 x HDMI | 1 x USB 1 YEAR
662    Ultra HD (4K) 3840 x 2160 Pixels 20 W Speaker ...
663    60 Hz Refresh Rate 2 x HDMI | 2 x USB 3 Years ...
664    16 Speaker Output 50 Hz Refresh Rate 2 x HDMI ...
665         50 Hz Refresh Rate 2 x HDMI | 2 x USB 1 YEAR
Name: pq-s-f, Length: 666, dtype: object

662

510

pandas.core.frame.DataFrame

,0
0,60
1,60
2,50
3,60
4,60
...,...
661,60
662,50
663,60
664,50


,0
0,True
1,True
2,False
3,True
4,True
...,...
661,True
662,False
663,True
664,False


0    510
dtype: int64

### Q02.

TV의 해상도(`Picture_quality`)는 HD, 4K, 8K로 나뉜다. 최근 시장에 등장한 8K 제품군은 상대적으로 비싸 비교적 매출이 떨어지는 편이다. 하지만 8K 제품군 구매자가 4K 제품군 구매자 대비 얼마나 만족하고 있는지 확인해보고자 한다.  
8K 제품군의 평가 평균 점수(`Stars`)의 평균값과 4K 제품군의 평가 평균 점수와의 평균값 차이의 절대값을 산출하시오.

※ 화면 해상도는 관련 변수에서 `"HD"`, `"4K"`, `"8K"`로 찾아낼 수 있다.  
※ 화면 해상도 정보는 `"Operating_system"`, `"channel"`, `"Picture_quality"` 변수에 흩어져 있다.  
※ 결과는 반올림하여 소수점 둘째 자리까지 출력하시오. `(정답 예시: 0.12)`

In [31]:
df_q2 = df.copy()
cols_q2 = ['channel', 'Operating_system', 'Picture_quality']
df_q2['c-o-pq'] = df_q2['channel'] + ' ' + df_q2['Operating_system'] + ' ' + df_q2['Picture_quality']
display(df_q2['c-o-pq'])

cond_q2_1 = df_q2['c-o-pq'].str.contains('8K')
stars_8k = df_q2.loc[cond_q2_1, ]['Stars'].mean()
display(cond_q2_1.sum(), stars_8k)

cond_q2_2 = df_q2['c-o-pq'].str.contains('4K')
stars_4k = df_q2.loc[cond_q2_2, ]['Stars'].mean()
display(cond_q2_2.sum(), stars_4k) 
display(round(abs(stars_8k-stars_4k), 2))

0      HD Ready 1366 x 768 Pixels 20 Speaker Output 6...
1      Netflix|Disney+Hotstar|Youtube Operating Syste...
2      Netflix|Prime Video|Disney+Hotstar|Youtube Ope...
3      Netflix|Prime Video|Disney+Hotstar|Youtube Ope...
4      Netflix|Prime Video|Disney+Hotstar|Youtube Ope...
                             ...                        
661    Ultra HD (4K) 3840 x 2160 pixels Pixels NO Spe...
662    Netflix|Prime Video|Youtube Operating System: ...
663    HD Ready 1366 x 786 Pixels 20 Speaker Output 6...
664    Netflix|Prime Video|Apple TV|Disney+Hotstar|Yo...
665    Ultra HD (4K) 3840 x 2160 pixels Pixels 10W + ...
Name: c-o-pq, Length: 666, dtype: object

1

3.6

365

3.2180821917808218

0.38

### Q03.

좋은 평가를 받는 제품의 조건을 알아보고자 한다. 이를 위해 기존 변수와 더불어 여러 파생변수를 생성 후 해당 변수들을 독립변수로 하고 평가 평균 점수(`Stars`)를 종속변수로 하여 Random Forest 분석을 통해 좋은 평가를 받는 제품의 조건을 알아보고자 한다.  
아래 절차를 수행하여 변수 중요도를 확인하고, 그 값 중 가장 큰 값을 가진 변수명을 기술하시오.

| 구분 | 변수 |
|---|---|
| 독립변수 | 후기 작성 비율: 후기 수(`Reviews`) / 평가 수(`Ratings`) |
| . | 공장 출고가(`MRP`) |
| . | 할인율: 현재 판매가(`current_price`) / 공장 출고가(`MRP`) |
| . | Netflix 제공 여부: 제공(1), 미제공(0) |
| . | Prime Video 제공 여부: 제공(1), 미제공(0) |
| . | 고해상도 여부: 4K 또는 8K(1), 나머지 0 |
| 종속변수 | 평가 평균 점수(`Stars`) |

※ `channel` 변수에 해상도(`'Pixel'` 문자 참조) 또는 운영체제 정보(`'Oper'` 문자 참조)가 있는 TV는 분석에서 제외하시오.  
※ Netflix와 Prime Video 제공여부 변수는 `channel` 변수를 참고하시오.  
※ 독립변수를 생성하면서 결측치가 생성되는 경우 해당 행을 모두 제거하시오.  
※ 학습 대상이 되는 행 개수는 197이다.  
※ 고해상도 여부 변수 생성은 해상도(`Picture_quality`) 변수만 참고하여 생성하시오.  
※ `seed`는 123으로 지정하시오. `(정답 예시: 고해상도여부 - 띄어쓰기 않고 문제에서 제시된 독립변수명을 기입)`

In [ ]:
df_q3 = df.copy()
df_q3['R/R'] = df_q3['Reviews'] / df_q3['Ratings']
df_q3['c/M'] = df_q3['current_price'] / df_q3['MRP']

display(df_q3['channel'].value_counts())
cond_q3_1 = df_q3['channel'].str.contains('Pixel')
cond_q3_2 = df_q3['channel'].str.contains('Oper')
cond_q3_3 = cond_q3_1 | cond_q3_2
display(cond_q3_1.sum(), cond_q3_2.sum(), cond_q3_3.sum())

df_q3_2 = df_q3.loc[~cond_q3_3, :].copy()
display(df_q3_2.shape)
cond_q3_4 = df_q3['channel'].str.contains('Netflix').astype(int)
df_q3_2['Netflix'] = cond_q3_4
cond_q3_5 = df_q3['channel'].str.contains('Prime Video').astype(int)
df_q3_2['PrimeVideo'] = cond_q3_5
display(df_q3_2.dtypes)

cond_q3_6 = df_q3_2['Picture_quality'].str.contains('4K|8K').astype(int) #and는 ?? #정규식검색에 함부러 띄어쓰기 X
df_q3_2['high_quality'] = cond_q3_6
display(df_q3_2['high_quality'].value_counts())

Netflix|Prime Video|Disney+Hotstar|Youtube             294
Netflix|Disney+Hotstar|Youtube                         171
HD Ready 1366 x 768 Pixels                              47
Netflix|Youtube                                         28
Netflix|Prime Video|Youtube                             21
Prime Video|Disney+Hotstar|Youtube                      18
Full HD 1920 x 1080 Pixels                              14
Ultra HD (4K) 3840 x 2160 Pixels                        13
Operating System: WebOS                                  6
Disney+Hotstar|Youtube                                   6
Operating System: Tizen                                  5
Netflix|Prime Video|Apple TV|Disney+Hotstar|Youtube      4
Youtube                                                  4
Operating System: Android                                3
HD Ready 1366*768 Pixels                                 3
HD Ready 1366 X 768 Pixels                               3
Ultra HD (4K) 3840 x 2160 pixels Pixels                 

99

19

118

(548, 13)

Product_Name         object
Stars               float64
Ratings               int64
Reviews               int64
current_price         int64
MRP                   int64
channel              object
Operating_system     object
Picture_quality      object
Speaker              object
Frequency            object
R/R                 float64
c/M                 float64
Netflix               int32
PrimeVideo            int32
dtype: object

1    329
0    219
Name: high_quality, dtype: int64

In [90]:
display(df_q3_2.isna().sum())
df_q3_na = df_q3_2.dropna().copy()
display(df_q3_na.shape)
#DNME
#D
cols_X = ['R/R', 'MRP', 'c/M', 'Netflix', 'PrimeVideo', 'high_quality']
X = df_q3_na[cols_X]
cols_y = ['Stars']
y = df_q3_na[cols_y]
display(X.dtypes, y.dtypes)

#N

#M
model = RandomForestRegressor(random_state = 123)
model.fit(X,y)
#y_pred = model.predict(X) ?????
#display(y_pred.shape) ????????
dict_ = dict(zip(cols_X, model.feature_importances_))
display(dict_)
ser_fi = pd.Series(dict_) #dataframe으로 만들지 말고 series로
display(ser_fi)
display(ser_fi.sort_values(ascending = False)) # sort() 도 있나?
display(ser_fi.max(),ser_fi.idxmax())

#E

Product_Name          0
Stars                 0
Ratings               0
Reviews               0
current_price         0
MRP                   0
channel               0
Operating_system      0
Picture_quality       0
Speaker               0
Frequency             0
R/R                 351
c/M                   0
Netflix               0
PrimeVideo            0
high_quality          0
dtype: int64

(197, 16)

R/R             float64
MRP               int64
c/M             float64
Netflix           int32
PrimeVideo        int32
high_quality      int32
dtype: object

Stars    float64
dtype: object

x:\study\docs\data_science\.venv\lib\site-packages\ipykernel_launcher.py:16: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  


{'R/R': 0.3122061680649266,
 'MRP': 0.17383962649966622,
 'c/M': 0.4446911950647382,
 'Netflix': 0.027140866837663043,
 'PrimeVideo': 0.02400441733105043,
 'high_quality': 0.01811772620195551}

R/R             0.312206
MRP             0.173840
c/M             0.444691
Netflix         0.027141
PrimeVideo      0.024004
high_quality    0.018118
dtype: float64

c/M             0.444691
R/R             0.312206
MRP             0.173840
Netflix         0.027141
PrimeVideo      0.024004
high_quality    0.018118
dtype: float64

0.4446911950647382

'c/M'

# 오답노트

## 최종 답

- Q01: 주사율이 60Hz인 TV는 **510개**
- Q02: 8K와 4K 제품군의 평균 평점 차이는 **0.38**
- Q03: 변수 중요도가 가장 큰 변수는 **할인율(`current_price / MRP`)**

## Q01 — 여러 열의 문자열에서 주사율 찾기

### 1. 문제에서 보장한 형식만 사용하기

문제는 주사율이 2~3자리 숫자 뒤에 `Hz` 단위로 표시된다고 보장한다. `Refresh Rate`까지 항상 존재한다고 가정하면 안 된다. 따라서 `Hz`를 기준으로 정규식을 작성한다.

### 2. 문자열 열을 합칠 때 공백 넣기

```python
df_q1['f-p-s'] = (
    df_q1['Frequency'] + ' ' +
    df_q1['Picture_quality'] + ' ' +
    df_q1['Speaker']
)
```

공백 없이 합치면 서로 다른 열의 마지막 문자와 첫 문자가 붙어 검색이 어려워질 수 있다.

### 3. `isin()`과 `str.contains()`의 차이

- `isin(['60 Hz'])`: 셀의 값 전체가 목록의 값과 같은지 검사한다.
- `str.contains('60 Hz')`: 셀의 문자열 일부에 해당 문자가 포함되어 있는지 검사한다.

이 문제처럼 긴 문자열 내부를 검색할 때는 `str.contains()` 또는 `str.extract()`를 사용한다.

### 4. `extract()`는 괄호 안의 부분만 반환한다

```python
freq = df_q1['f-p-s'].str.extract(
    r'(?P<Frequency>\d{2,3})\s*Hz',
    expand=False
)
answer = (freq == '60').sum()
```

- `\d{2,3}`: 숫자 2~3자리
- `\s*`: 공백 0개 이상
- `Hz`: 숫자 뒤에 반드시 있어야 하는 검색 조건
- `(?P<Frequency>...)`: 추출할 그룹의 열 이름을 `Frequency`로 지정
- `expand=False`: 결과를 1차원 Series로 반환

괄호 안에는 숫자만 있으므로 `Hz`는 검색 조건으로 사용될 뿐 결과에는 포함되지 않는다. `Hz`까지 추출하려면 `r'(?P<Frequency>\d{2,3}\s*Hz)'`처럼 `Hz`도 괄호 안에 넣는다.

`expand`를 생략하면 기본값이 `True`이므로 결과가 1열짜리 DataFrame이다. 이 경우 `(freq == '60').sum()`도 숫자 하나가 아니라 Series를 반환한다. 개수 하나만 필요하다면 `expand=False`가 편하다.

추출된 값은 문자열이므로 숫자 `60`이 아닌 문자열 `'60'`과 비교해야 한다. 숫자로 비교하려면 먼저 `pd.to_numeric()`으로 변환한다.

#### 복사본 풀이에서 확인한 DataFrame 열 이름 처리

복사본에서는 `expand=True`인 기본 설정으로 추출하여 결과가 DataFrame이고 열 이름이 정수 `0`이 되었다. 다음 세 방법 중 하나로 명확하게 처리할 수 있다.

```python
# 방법 1: 이름 있는 캡처 그룹 사용
freq_df = df_q1['f-p-s'].str.extract(r'(?P<freq>\d{2,3})\s*Hz')
answer = freq_df['freq'].eq('60').sum()

# 방법 2: 추출 후 열 이름 변경
freq_df = df_q1['f-p-s'].str.extract(r'(\d{2,3})\s*Hz')
freq_df = freq_df.rename(columns={0: 'freq'})
answer = freq_df['freq'].eq('60').sum()

# 방법 3: 처음부터 Series로 반환
freq = df_q1['f-p-s'].str.extract(r'(\d{2,3})\s*Hz', expand=False)
answer = freq.eq('60').sum()
```

`freq_df.isin(['60']).sum()`도 계산은 되지만 열별 합계를 담은 Series를 반환한다. 최종 답 하나를 구할 때는 특정 열을 선택한 뒤 `.eq('60').sum()`을 쓰는 편이 결과 형태가 분명하다.

### 5. 개수만 구할 때는 `contains()`도 가능하다

```python
answer = df_q1['f-p-s'].str.contains(
    r'(?<!\d)60\s*Hz\b',
    na=False
).sum()
```

단순히 `'60 Hz'`만 검색하면 `60Hz`처럼 공백이 없는 표기를 놓친다. 숫자 경계를 지정하지 않으면 `160 Hz`에도 `60 Hz`가 포함될 수 있으므로 정규식으로 정확한 60Hz를 검사하는 편이 안전하다.

## Q02 — OR 조건과 평균 차이

`Operating_system`, `channel`, `Picture_quality`를 공백과 함께 합친 뒤 각각 `8K`, `4K` 포함 여부를 검사한 접근은 맞다. 두 평균의 차이는 방향이 아닌 크기를 묻기 때문에 `abs()`를 사용하고 마지막에 소수점 둘째 자리까지 반올림한다.

```python
answer = round(abs(stars_8k - stars_4k), 2)
```

## Q03 — 조건 결합, 결측치, 모델 입력 형태

### 1. `str.contains()`의 OR, AND, NOT

```python
# OR: Pixel 또는 Oper가 포함됨
cond_or = (
    df_q3['channel'].str.contains('Pixel', na=False)
    | df_q3['channel'].str.contains('Oper', na=False)
)

# 정규식 하나로 표현한 OR
cond_or = df_q3['channel'].str.contains(r'Pixel|Oper', na=False)

# AND: Netflix와 Youtube가 모두 포함됨
cond_and = (
    df_q3['channel'].str.contains('Netflix', na=False)
    & df_q3['channel'].str.contains('Youtube', na=False)
)

# NOT: 조건에 해당하지 않음
df_filtered = df_q3.loc[~cond_or].copy()
```

Series 조건에서는 파이썬의 `or`, `and`, `not`이 아니라 `|`, `&`, `~`를 사용하고 각 조건을 괄호로 감싼다. 정규식에서 `4K|8K`의 `|`도 OR이다. AND를 정규식 문자열에 단순히 띄어 써서 표현할 수는 없다.

### 2. 파생변수는 필터링된 DataFrame에서 만들기

기존 코드처럼 원본에서 만든 Series를 필터링된 DataFrame에 넣어도 인덱스 정렬 덕분에 현재 결과는 맞다. 그러나 혼동을 줄이려면 필터링한 `df_q3_2`에서 직접 만드는 것이 명확하다.

```python
df_q3_2['Netflix'] = df_q3_2['channel'].str.contains('Netflix', na=False).astype(int)
df_q3_2['PrimeVideo'] = df_q3_2['channel'].str.contains('Prime Video', na=False).astype(int)
df_q3_2['high_quality'] = df_q3_2['Picture_quality'].str.contains(
    r'4K|8K', na=False
).astype(int)
```

`str.contains()`의 결과는 Boolean이므로 `.astype(int)`를 적용하면 `True → 1`, `False → 0`인 이진 변수가 된다.

### 3. `isna()` 뒤에 사용하는 집계

- `df.isna().sum()`: 열별 결측치 개수
- `df.isna().any()`: 열마다 결측치가 하나라도 있는지 확인
- `df.isna().all()`: 열의 값이 모두 결측치인지 확인
- `df.isna().any(axis=1)`: 행마다 결측치가 하나라도 있는지 확인
- `df.isna().all(axis=1)`: 행의 값이 모두 결측치인지 확인

이 문제에서는 `Ratings`가 0이면 `Reviews / Ratings`의 결과가 유효하지 않아 결측치가 생길 수 있다. 입력변수를 만든 다음 결측치 행을 제거하며, 최종 학습 데이터가 문제에서 제시한 **197개**인지 확인한다.

전체 열을 기준으로 한 `dropna()`도 현재 데이터에서는 197개가 되어 정답이지만, 모델에 사용하지 않는 열의 결측치까지 제거할 수 있다. 일반적으로는 입력변수와 종속변수만 지정하는 편이 안전하다.

```python
model_cols = cols_X + ['Stars']
df_q3_na = df_q3_2.dropna(subset=model_cols).copy()
```

### 4. `y`는 1차원 Series로 전달하기

```python
X = df_q3_na[cols_X]       # DataFrame: (197, 6)
y = df_q3_na['Stars']      # Series: (197,)

model = RandomForestRegressor(random_state=123)
model.fit(X, y)
```

`df_q3_na[['Stars']]`처럼 대괄호를 두 번 쓰면 `(197, 1)` 형태의 2차원 DataFrame이 된다. 단일 종속변수를 기대하는 모델에 이를 전달하면 `A column-vector y was passed when a 1d array was expected` 경고가 발생한다. 대괄호를 한 번 사용하면 경고가 해결된다. 이미 2차원으로 만들었다면 `y.squeeze()` 또는 `y.values.ravel()`도 가능하다.

### 5. 변수 중요도 확인

```python
importance = pd.Series(
    model.feature_importances_,
    index=cols_X
).sort_values(ascending=False)

display(importance)
display(importance.idxmax(), importance.max())
```

Series를 실제로 정렬하는 메서드는 `sort()`가 아니라 `sort_values()`이다. 가장 중요한 변수의 이름은 `idxmax()`, 중요도 값은 `max()`로 구한다. 이 문제는 예측값을 요구하지 않으므로 `model.predict(X)`는 필요 없다.

## 기억할 핵심

1. 긴 문자열 내부 검색은 `str.contains()`, 일부 값 추출은 `str.extract()`를 사용한다.
2. `extract()`는 캡처 그룹 `()` 안의 값만 반환하며, `expand=False`이면 Series가 된다.
3. Series 조건 결합은 `|`, `&`, `~`를 사용한다.
4. 결측치는 `sum()`으로 개수, `any()`로 하나라도 존재하는지, `all()`로 전부 결측인지 확인한다.
5. 회귀모델의 단일 `y`는 `df['Stars']`처럼 1차원으로 전달한다.